src/meta/src/stream/stream_graph/schedule.rs

```diff
@@ -224,15 +294,27 @@ impl Scheduler {
         if !failed.is_empty() {
             bail!("Failed to schedule: {:?}", failed);
         }
-        assert_eq!(success.len(), graph.graph.fragments.len());
+        // Should not contain any existing fragments.
+        assert_eq!(success.len(), graph.building_fragments().len());
 
         // Extract the results.
         let distributions = success
             .into_iter()
-            .map(|Success(id, distribution)| {
-                let distribution = match distribution {
-                    DistId::Hash(mapping) => Distribution::Hash(all_hash_mappings[mapping].clone()),
-                    DistId::Singleton => Distribution::Singleton,
+            .map(|Success(id, result)| {
+                let distribution = match result {
+                    // Required
+                    Result::Required(DistId::Singleton(parallel_unit)) => {
+                        Distribution::Singleton(parallel_unit)
+                    }
+                    Result::Required(DistId::Hash(mapping)) => {
+                        Distribution::Hash(all_hash_mappings[mapping].clone())
+                    }
+
+                    // Default
+                    Result::DefaultSingleton => {
+                        Distribution::Singleton(self.default_singleton_parallel_unit)
+                    }
```

Comment:
Should we use different parallel unit for a streaming job with multiple singleton fragments instead of a fixed one? For example `create materialized view v as select count(*) from t1 union all select count(*) from t2 union all select count(*) from t3;` this sql has 3 singleton fragments.

Reply:
Good catch. I've also thought of this but I'm still trying to find a way to represent it in `crepe`, as the current rules only propagate the _requirement_ through no-shuffle, but don't go back and check whether the final result is still correct. If we assign different parallel units to them, it's possible that some no-shuffles are violated... Maybe we need a `SameDist` relation?

Reply:
It seems like a tough task to assign different parallel units for different singleton fragments with the `NoShuffle` restrictions, because when we assign a parallel unit to a singleton fragment, we need to propagate it to other fragment connected by `NoShuffle` edge and after that we can assign another different parallel unit to another singleton fragment, but `crepe` might can not guarantee this assign ordering. It could assign two different parallel units before propagation and then violate the `NoShuffle` restrictions.

Reply:
Or we can use a `crepe` schedule loop (instead of one `crepe` schedule) to assign a parallel unit to `Result::DefaultSingleton` and propagate it through the `NoShuffle` edge. Once all fragments get their specific parallel unit we are done, but it would be more complicated.

Reply:
Agree. For now, assuming that the singleton executors won't process too much data and there won't be too many singletons in a graph, the current implementation just works.

As long as we extract the scheduling of the whole graph as a separate step like this, we can benefit from this global knowledge of scheduling and make the following steps easier to implement. The algorithm can be optimized or replaced if necessary in the future.